WiseUnpacker/SabreTools.Serialization.Wrappers/WiseSectionHeader.cs

```diff
@@ -214,49 +218,197 @@ private bool ExtractHeaderDefinedFiles(Stream data, string outputDirectory, bool
+            if (ExtractFile(data, "FirstExecutable.exe", outputDirectory, FirstExecutableFileEntryLength, includeDebug)
+                != ExtractionStatus.GOOD)
                 return false;
-            data.Seek(data.Position - MsiFileEntryLength, 0);
 
             // Extract second executable, if it exists
-            expected = new DeflateInfo
-            {
-                InputSize = SecondExecutableFileEntryLength,
-                OutputSize =
-                    SecondExecutableFileEntryLength - 4,
-                Crc32 = 0
-            };
-            data.Seek(data.Position - SecondExecutableFileEntryLength, 0);
-            if (InflateWrapper.ExtractFile(data, "WiseScript.bin", outputDirectory, expected, false, includeDebug)
-                == ExtractionStatus.FAIL)
+            if (ExtractFile(data, "SecondExecutable.exe", outputDirectory, SecondExecutableFileEntryLength, 
+                includeDebug)
+                != ExtractionStatus.GOOD)
+                return false;
+            
+            // Extract second executable, if it exists
+            // If there's a size provided for the second executable but no size for the first executable, the size of
+            // the second executable appears to be some unrelated value that's larger than the second executable
+            // actually is. Currently unable to extract properly in these cases, as no header value in such installers
+            // seems to actually correspond to the real size of the second executable.
+            if (ExtractFile(data, "ThirdExecutable.exe", outputDirectory, ThirdExecutableFileEntryLength, 
+                    includeDebug)
+                != ExtractionStatus.GOOD)
                 return false;
-            data.Seek(data.Position - SecondExecutableFileEntryLength, 0);
+            
+            // Extract main MSI file
+            if (ExtractFile(data, "ExtractedMsi.msi", outputDirectory, MsiFileEntryLength, includeDebug)
+                != ExtractionStatus.GOOD)
+                return false;
+            
+            return true;
+        }
 
-            // Extract first executable, if it exists
-            expected = new DeflateInfo
+        /// <summary>
+        /// Attempt to extract a file defined by a filename
+        /// </summary>
+        /// <param name="source">Stream representing the deflated data</param>
+        /// <param name="filename">Output filename, null to auto-generate</param>
+        /// <param name="outputDirectory">Output directory to write to</param>
+        /// <param name="entrySize">Expected size of the file plus crc32</param>
+        /// <param name="includeDebug">True to include debug data, false otherwise</param>
+        /// <returns>Extraction status representing the final state</returns>
+        /// <remarks>Assumes that the current stream position is the end of where the data lives</remarks>
+        private ExtractionStatus ExtractFile(Stream source,
```

Comment:
I'm aware I stated that it's okay to have duplicate code to get things working. Is it possible to use the existing extraction at this point?